core/math/bvh_cull.inc

```diff
@@ -138,8 +141,14 @@ void _cull_hit(uint32_t p_ref_id, CullParams &p) {
 	if (USE_PAIRS) {
 		const ItemExtra &ex = _extra[p_ref_id];
 
-		if (!(p.mask & ex.pairable_type))
-			return;
+		// double check this as a possible source of bugs in future.
+		bool from_match_to = p.mask & ex.pairable_type;
+
+		if (!from_match_to) {
+			bool to_match_from = ex.pairable_mask & p.pairable_type;
+			if (!to_match_from)
+				return;
+		}
```

Comment:
@reduz I could do with you double checking this. I've never been super sure of how the masks are intended to work.

These are the `pairable_mask` and `pairable_type` as passed into the octree. So here we are doing a final check from the caller `p` (e.g. object A) and the collidee `ex` (e.g. object B).

If neither matches it is ignoring the collision.

Reply:
Ok have confirmed this with reduz, we think this is correct now with the masks. :+1: